### PR TITLE
MapOperator and JoinOperator Implementation.

### PR DESCRIPTION
**Pull request by sukritkalra**
- This PR provides the implementation of MapOperator and JoinOperator along with the relevant tests.

### CODE REVIEW

File: src/dataflow/operators/join_operator.rs [hunk 1]
```diff
@@ -1,83 +1,179 @@
...
+        // Get the vector corresponding to the given key and acquire a write lock on it.
+        let mut messages = _msgs.get_mut(timestamp).unwrap();
+        messages.push(msg);
+        // The data structure automatically releases the write lock once the guard goes out of
+        // scope.
+    }
+
+    /// Cleans the state corresponding to a given Timestamp (upto and including).
+    fn clean_state(&self, timestamp: &Timestamp) {
+        let _timestamps = &mut (*(*self._timestamps).write().unwrap());
```
Comment (pschafhalter) [inline src/dataflow/operators/join_operator.rs]:
```suggestion
        let timestamps = self.timestamps.write().unwrap();
```

File: src/dataflow/operators/join_operator.rs [hunk 2]
```diff
@@ -1,83 +1,179 @@
...
+        while _timestamps.len() > 0 && _timestamps.peek().unwrap().0 <= *timestamp {
```
Comment (pschafhalter) [inline src/dataflow/operators/join_operator.rs]: Might need some small fixes but this  should work:
```suggestion
        while timestamps.peek().map_or(false, |t| t <= timestamp) {
```

File: src/dataflow/operators/join_operator.rs [hunk 3]
```diff
@@ -1,83 +1,179 @@
...
+            let _t = _timestamps.pop().unwrap().0;
```
Comment (pschafhalter) [inline src/dataflow/operators/join_operator.rs]:
```suggestion
            let t = timestamps.pop().unwrap().0;
```

File: src/dataflow/operators/join_operator.rs [hunk 4]
```diff
@@ -1,83 +1,179 @@
...
+            (*(self.msgs.write().unwrap())).remove(&_t).unwrap();
```
Comment (pschafhalter) [inline src/dataflow/operators/join_operator.rs]:
```suggestion
            self.msgs.write().unwrap().remove(&t).expect("WindowState: expected timestamp {:?} to be present in self.msgs");
```

File: src/dataflow/operators/join_operator.rs [hunk 5]
```diff
@@ -1,83 +1,179 @@
...
+        }
+    }
+
+    /// Retrieve the state.
+    fn get_state(&self, timestamp: &Timestamp) -> Vec<D> {
```
Comment (pschafhalter) [inline src/dataflow/operators/join_operator.rs]:
```suggestion
    fn get_state(&self, timestamp: &Timestamp) -> Option<Vec<D>> {
```

File: src/dataflow/operators/join_operator.rs [hunk 6]
```diff
@@ -1,83 +1,179 @@
...
+        (*((*(self.msgs.read().unwrap())).get(timestamp).unwrap())).clone()
```
Comment (pschafhalter) [inline src/dataflow/operators/join_operator.rs]:
```suggestion
        self.msgs.read().unwrap().get(timestamp).cloned()
```

File: src/dataflow/operators/join_operator.rs [hunk 7]
```diff
@@ -1,83 +1,179 @@
...
+        // Package the state with the left stream and add a callback to the new stream.
+        let stateful_stream_left = input_stream_left.add_state(StreamState::<D1>::new());
+        stateful_stream_left.add_callback(Self::on_left_data_callback);
+
+        // Package the state with the right stream and add a callback to the new stream.
+        let stateful_stream_right = input_stream_right.add_state(StreamState::<D2>::new());
+        stateful_stream_right.add_callback(Self::on_right_data_callback);
+
+        let cb = config.arg.unwrap();
+        stateful_stream_left
```
Comment (pschafhalter) [inline src/dataflow/operators/join_operator.rs]: Could be done with `add_watermark_callback!` macro, but this should be fine.

File: src/dataflow/operator.rs [hunk 8]
```diff
@@ -32,7 +32,7 @@ impl<T: Clone> OperatorConfig<T> {
     pub fn new(name: &str, arg: T, flow_watermarks: bool, node_id: NodeId) -> Self {
         Self {
             name: name.to_string(),
-            id: OperatorId::nil(),
+            id: OperatorId::new_deterministic(),
```
[the remark this answers is not in the document]
Comment (sukritkalra) [reply src/dataflow/operator.rs]: I think this is from the changes in the master. We have `OperatorId::nil()` now. This change seems to be outdated so I'll resolve this.

File: src/dataflow/operators/join_operator.rs [hunk 9]
```diff
@@ -1,88 +1,178 @@
...
 
-pub struct JoinOperator<D1: Data, D2: Data, D3: Data> {
-    name: String,
-    id: OperatorId,
-    phantom: PhantomData<(D1, D2, D3)>,
-}
+    /// Adds a message to the ConcurrentHashMap.
+    fn add_msg(&mut self, timestamp: &Timestamp, msg: D) {
+        // Insert a new Vector if the key does not exist, and add the key to the timestamps.
+        let _msgs = &(*(*self.msgs).write().unwrap());
```
[the remark this answers is not in the document]
Comment (sukritkalra) [reply src/dataflow/operators/join_operator.rs]: Great! This looks better, thanks!

**Pull request merged**